File: 3534.path-existence-queries-in-a-graph-ii.py
#
# @lc app=leetcode id=3534 lang=python3
#
# [3534] Path Existence Queries in a Graph II
#

from typing import List

# @lc code=start
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        sorted_nodes = sorted(range(n), key=lambda x: nums[x])
        pos = [0] * n
        for r, node in enumerate(sorted_nodes):
            pos[node] = r
        
        # Compute right_reach
        right_reach = [0] * n
        j = 0
        for i in range(n):
            while j < n and nums[sorted_nodes[j]] <= nums[sorted_nodes[i]] + maxDiff:
                j += 1
            right_reach[i] = j - 1
        
        # Compute comp_rank
        comp_rank = [0] * n
        cid = 0
        comp_rank[0] = 0
        for i in range(1, n):
            if nums[sorted_nodes[i]] - nums[sorted_nodes[i - 1]] > maxDiff:
                cid += 1
            comp_rank[i] = cid
        
        # Binary lifting for right_reach iterations
        LOG = 18
        jump = [[0] * n for _ in range(LOG)]
        for i in range(n):
            jump[0][i] = right_reach[i]
        for k in range(1, LOG):
            for i in range(n):
                mid = jump[k - 1][i]
                jump[k][i] = jump[k - 1][mid]
        
        ans = []
        for u, v in queries:
            if u == v:
                ans.append(0)
                continue
            p1 = pos[u]
            p2 = pos[v]
            if p1 > p2:
                p1, p2 = p2, p1
            if comp_rank[p1] != comp_rank[p2]:
                ans.append(-1)
                continue
            # Binary search minimal m
            l, h = 1, n
            while l < h:
                m = (l + h) // 2
                cur = p1
                reached = False
                for k in range(LOG - 1, -1, -1):
                    if (1 << k) & m:
                        cur = jump[k][cur]
                        if cur >= p2:
                            reached = True
                            break
                if reached or cur >= p2:
                    h = m
                else:
                    l = m + 1
            ans.append(l)
        return ans
# @lc code=end

File: 3534.path-existence-queries-in-a-graph-ii.cpp
#
# @lc app=leetcode id=3534 lang=cpp
#
# [3534] Path Existence Queries in a Graph II
#
# @lc code=start
class Solution {
public:
    vector<int> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        // Build adjacency list
        vector<vector<int>> adj(n);
        
        // Sort nodes by their values to efficiently find edges
        vector<pair<int, int>> sorted_nodes;
        for (int i = 0; i < n; i++) {
            sorted_nodes.push_back({nums[i], i});
        }
        sort(sorted_nodes.begin(), sorted_nodes.end());
        
        // For each node, find all nodes within maxDiff
        for (int i = 0; i < n; i++) {
            int idx1 = sorted_nodes[i].second;
            int val1 = sorted_nodes[i].first;
            
            for (int j = i + 1; j < n; j++) {
                int idx2 = sorted_nodes[j].second;
                int val2 = sorted_nodes[j].first;
                
                if (val2 - val1 > maxDiff) break;
                
                adj[idx1].push_back(idx2);
                adj[idx2].push_back(idx1);
            }
        }
        
        // Process queries
        vector<int> result;
        for (const auto& query : queries) {
            int u = query[0];
            int v = query[1];
            
            if (u == v) {
                result.push_back(0);
                continue;
            }
            
            // BFS to find shortest path
            queue<int> q;
            vector<int> dist(n, -1);
            q.push(u);
            dist[u] = 0;
            
            while (!q.empty()) {
                int curr = q.front();
                q.pop();
                
                if (curr == v) break;
                
                for (int neighbor : adj[curr]) {
                    if (dist[neighbor] == -1) {
                        dist[neighbor] = dist[curr] + 1;
                        q.push(neighbor);
                    }
                }
            }
            
            result.push_back(dist[v]);
        }
        
        return result;
    }
};
# @lc code=end